docs/pallas/tpu/core_map.md

```diff
@@ -0,0 +1,357 @@
+  def _():
+    pl.run_scoped(
+        partial(swap_cores_kernel, x_hbm_ref, o_hbm_ref),
+        *([pltpu.VMEM(local_vmem_shape, x.dtype)] * 3),  # VMEM allocations
+        *([pltpu.SemaphoreType.DMA] * 3),                # semaphores
+    )
+  return o_hbm_ref[...]
+
+
+x = jax.random.normal(jax.random.key(0), input_shape, jnp.float32)
+x = jax.device_put(x, sharding)
+y = swap_cores(x)
+
+np.testing.assert_array_equal(y[:, 128:], x[:, :128] * 2)
+np.testing.assert_array_equal(y[:, :128], x[:, 128:] * 2)
+```
+
+### Save the boilerplate
+
+For `jax>=0.8.1`, you can use decorator `pl.kernel` to wrap boilerplates such as `core_map`, `run_scoped`, and output buffer allocation.
+
+Note that this should run inside any `jax.shard_map` you may have at the top level.
+
+
+```python
+@jax.jit
+@partial(jax.shard_map, mesh=mesh, in_specs=in_spec, out_specs=in_spec, check_vma=False)
+def swap_cores(x):
+  scratch_shapes = [pltpu.VMEM(local_vmem_shape, x.dtype)] * 3 + [pltpu.SemaphoreType.DMA] * 3
+  return pl.kernel(swap_cores_kernel, out_shape=x, mesh=tc_mesh,
+                   scratch_shapes=scratch_shapes,
+                   compiler_params=pltpu.CompilerParams(collective_id=0))(x)
+
+y = swap_cores(x)
+np.testing.assert_array_equal(y[:, 128:], x[:, :128] * 2)
+np.testing.assert_array_equal(y[:, :128], x[:, 128:] * 2)
+```
+
+## Pipelining with `core_map`
+
+Note that the kernel above only does simple copies and compute, without automatic pipelining via Pallas `grid` and `BlockSpec`. To do pipelining inside `core_map`, use `pltpu.emit_pipeline` inside the core-local kernel.
+
+**Automatically parallelize work amongst cores**
+
+The simple way is to annotate a block axis as "core parallelizable", and Pallas will automatically parallelize the work under the hood. Both `pl.pallas_call` and `pltpu.emit_pipeline` supports this, via arguments `core_axis` and `dimension_semantics`. The `pallas_call` example was [in another guide](https://docs.jax.dev/en/latest/pallas/tpu/pipelining.html#tpus-in-megacore-configuration), and the `emit_pipeline` case is shown below.
+
+**Scratch shapes allocation**
+
+Note that in the example below, the top level `pl.run_scoped` (wrapped inside `kernel`) did not allocate any scratch buffer. Instead, the VMEM scratch shapes were allocated within `pltpu.emit_pipeline`, which called another `pl.run_scoped` within.
+
+
+
+```python
+def add_one_body(in_vmem, out_vmem):
+  out_vmem[...] = in_vmem[...] + 1
+
+input_shape = (1024, 1024)
+in_spec = jax.P('device', None)
+
+def add_one_kernel(x_hbm_ref, o_hbm_ref):
+  in_shape = x_hbm_ref.shape
+  pltpu.emit_pipeline(
+      add_one_body,
+      grid=(in_shape[0] // 8, in_shape[1] // 128),
+      in_specs=[pl.BlockSpec(
+          block_shape=(8, 128), index_map=lambda i, j: (i, j),
+      )],
+      out_specs=[pl.BlockSpec(
+          block_shape=(8, 128), index_map=lambda i, j: (i, j),
+      )],
+      core_axis=0,
```

Comment:
This should be the string name of the core_axis in tc_mesh

Reply:
Chaanged to `core_axis_name='core'`